sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py

```diff
@@ -195,16 +195,65 @@ def test_baseline_categorical_crossentropy():
     loss = _LOSSES['categorical_crossentropy']()
     for y_train in (np.zeros(shape=100), np.ones(shape=100)):
         y_train = y_train.astype(np.float64)
-        baseline_prediction = loss.get_baseline_prediction(y_train,
+        baseline_prediction = loss.get_baseline_prediction(y_train, None,
                                                            prediction_dim)
         assert baseline_prediction.dtype == y_train.dtype
         assert_all_finite(baseline_prediction)
 
     # Same logic as for above test. Here inverse_link_function = softmax and
     # link_function = log
     y_train = rng.randint(0, prediction_dim + 1, size=100).astype(np.float32)
-    baseline_prediction = loss.get_baseline_prediction(y_train, prediction_dim)
+    baseline_prediction = loss.get_baseline_prediction(y_train, None,
+                                                       prediction_dim)
     assert baseline_prediction.shape == (prediction_dim, 1)
     for k in range(prediction_dim):
         p = (y_train == k).mean()
         assert np.allclose(baseline_prediction[k, :], np.log(p))
+
+
+@pytest.mark.parametrize('loss, problem', [
+    ('least_squares', 'regression'),
+    ('least_absolute_deviation', 'regression'),
+    ('binary_crossentropy', 'classification'),
+    ('categorical_crossentropy', 'classification')
+    ])
+def test_sample_weight(loss, problem):
+    rng = np.random.RandomState(42)
+    n_samples = 100
+    n_classes = 3
+    prediction_dim = n_classes if loss == "categorical_crossentropy" else 1
+    if problem == 'regression':
+        y_true = rng.normal(size=n_samples).astype(Y_DTYPE)
+    else:
+        y_true = rng.randint(0, n_classes, size=n_samples).astype(Y_DTYPE)
+    raw_predictions = rng.normal(
+        size=(prediction_dim, n_samples)
+    ).astype(Y_DTYPE)
+
+    sample_weight = rng.random(size=(n_samples,))
+    ones = np.ones(shape=(n_samples,))
+
+    loss_ = _LOSSES[loss]()
+    gradients0, hessians0 = loss_.init_gradients_and_hessians(
+        n_samples, prediction_dim, None)
+    loss_.update_gradients_and_hessians(gradients0, hessians0, y_true,
+                                        raw_predictions, None)
+
+    loss_ = _LOSSES[loss]()
+    gradients1, hessians1 = loss_.init_gradients_and_hessians(
+        n_samples, prediction_dim, ones)
+    loss_.update_gradients_and_hessians(gradients1, hessians1, y_true,
+                                        raw_predictions, ones)
+
+    # passing ones as sample weights shouldn't change the values
+    assert np.allclose(gradients0, gradients1)
+    assert np.allclose(hessians0, hessians1)
+
+    loss_ = _LOSSES[loss]()
+    gradients2, hessians2 = loss_.init_gradients_and_hessians(
+        n_samples, prediction_dim, ones)
+    loss_.update_gradients_and_hessians(gradients2, hessians2, y_true,
+                                        raw_predictions, sample_weight)
+
+    assert np.allclose(gradients1 * sample_weight, gradients2)
```

Comment:
@NicolasHug these two are equal, they're close.

Reply:
if you cast before multiplying to simulate what happens in Cython, are they equal?

Reply:
I thought they would be, but they're not.

``` python
np.r_[gradients1.astype(np.float32) * sample_weight.astype(np.float32), 
    gradients2.astype(np.float32)].T

array([[-1.67761183e+00, -1.67761183e+00],
       [-2.09178999e-01, -2.09178999e-01],
       [-6.90326512e-01, -6.90326512e-01],
       [-1.63349128e+00, -1.63349128e+00],
       [ 2.61952784e-02,  2.61952803e-02],
       [ 1.87366754e-01,  1.87366739e-01],
       [ 2.48452097e-01,  2.48452097e-01],
       [-4.80281323e-01, -4.80281323e-01],
       [ 6.30383074e-01,  6.30383074e-01],
       [-5.63474834e-01, -5.63474834e-01],
       [-7.44183958e-01, -7.44183958e-01],
       [ 2.20273912e-01,  2.20273927e-01],
       [-1.45075828e-01, -1.45075828e-01],
       [ 2.84458184e+00,  2.84458160e+00],
       [ 1.07580423e+00,  1.07580411e+00],
       [ 6.87433481e-01,  6.87433481e-01],
       [ 8.70531499e-01,  8.70531440e-01],
       [-5.01221657e-01, -5.01221597e-01],
       [ 7.70263135e-01,  7.70263135e-01],
       [ 2.03399166e-01,  2.03399166e-01],
       [-3.90117526e-01, -3.90117496e-01],
       [-2.45705396e-02, -2.45705396e-02],
       [ 6.21697247e-01,  6.21697247e-01],
       [ 1.24250017e-02,  1.24250008e-02],
       [ 3.24146867e-01,  3.24146867e-01],
       [ 1.22865725e+00,  1.22865725e+00],
       [ 4.89398651e-03,  4.89398651e-03],
       [-3.51818353e-02, -3.51818353e-02],
       [ 5.76058984e-01,  5.76058984e-01],
       [-7.62818605e-02, -7.62818605e-02],
       [-1.20574951e-01, -1.20574936e-01],
       [-9.31533277e-01, -9.31533217e-01],
       [-8.07574272e-01, -8.07574213e-01],
       [ 3.30487490e-01,  3.30487460e-01],
       [-1.08505595e+00, -1.08505595e+00],
       [ 2.36478880e-01,  2.36478880e-01],
       [-5.12743071e-02, -5.12743071e-02],
       [ 8.70150924e-01,  8.70150924e-01],
       [ 1.15788007e+00,  1.15787995e+00],
       [-9.10074949e-01, -9.10074949e-01],
       [-3.71037513e-01, -3.71037483e-01],
       [ 1.10834789e+00,  1.10834789e+00],
       [-7.70234883e-01, -7.70234883e-01],
       [ 1.56872094e-01,  1.56872094e-01],
       [ 1.38235545e+00,  1.38235545e+00],
       [ 4.06699836e-01,  4.06699866e-01],
       [-3.40778768e-01, -3.40778768e-01],
       [-1.86536223e-01, -1.86536238e-01],
       [ 4.52062767e-03,  4.52062767e-03],
       [ 1.98307955e+00,  1.98307967e+00],
       [-6.15207143e-02, -6.15207143e-02],
       [ 5.09126365e-01,  5.09126365e-01],
       [-1.26886694e-03, -1.26886705e-03],
       [-6.57517835e-02, -6.57517835e-02],
       [-1.15439117e-01, -1.15439124e-01],
       [-4.11807597e-01, -4.11807597e-01],
       [ 1.48565376e+00,  1.48565376e+00],
       [ 5.59560955e-01,  5.59560955e-01],
       [-1.00519466e+00, -1.00519466e+00],
       [-8.92965347e-02, -8.92965421e-02],
       [-4.73142833e-01, -4.73142833e-01],
       [ 7.17784464e-01,  7.17784464e-01],
       [ 1.25557339e+00,  1.25557339e+00],
       [ 2.29716018e-01,  2.29716003e-01],
       [ 6.32968023e-02,  6.32967949e-02],
       [-2.33724058e-01, -2.33724058e-01],
       [ 3.18264574e-01,  3.18264604e-01],
       [ 6.76953673e-01,  6.76953673e-01],
       [-8.73719528e-03, -8.73719528e-03],
       [-1.26073938e-02, -1.26073938e-02],
       [-5.75451665e-02, -5.75451665e-02],
       [-9.58693624e-02, -9.58693624e-02],
       [-3.53097096e-02, -3.53097096e-02],
       [-8.60919535e-01, -8.60919535e-01],
       [ 1.37341416e+00,  1.37341416e+00],
       [ 5.16637228e-04,  5.16637228e-04],
       [-3.64017822e-02, -3.64017859e-02],
       [ 8.29778850e-01,  8.29778910e-01],
       [-6.17321953e-02, -6.17321953e-02],
       [ 2.04245996e+00,  2.04245996e+00],
       [ 3.36871684e-01,  3.36871684e-01],
       [-7.47808337e-01, -7.47808397e-01],
       [-1.61871815e+00, -1.61871815e+00],
       [ 4.53376584e-02,  4.53376547e-02],
       [ 2.19159931e-01,  2.19159931e-01],
       [ 7.60893881e-01,  7.60893881e-01],
       [-2.22468987e-01, -2.22468987e-01],
       [-3.43949229e-01, -3.43949199e-01],
       [-2.08827943e-01, -2.08827958e-01],
       [-3.30449671e-01, -3.30449700e-01],
       [-3.83606441e-02, -3.83606404e-02],
       [-7.21091181e-02, -7.21091181e-02],
       [ 2.42882520e-02,  2.42882539e-02],
       [-5.37786841e-01, -5.37786901e-01],
       [ 5.31501889e-01,  5.31501889e-01],
       [ 1.06395531e+00,  1.06395531e+00],
       [-4.58031833e-01, -4.58031863e-01],
       [-6.90442249e-02, -6.90442324e-02],
       [ 2.43310574e-02,  2.43310574e-02],
       [-4.95629132e-01, -4.95629102e-01]], dtype=float32)
```

Reply:
The difference seems to be of the order of 1e-6, it's definitely reasonable IMO